README.md
# StudyZoom (VOIP Application)
Group Study Video calling application 

# Table of Contents

1. [Starting the app in Docker](#starting-the-app-in-docker)
2. [How to use it](#how-to-use-it)
   - [Set Up for Unix, MacOS](#set-up-for-unix-macos)
   - [Set Up for Windows](#set-up-for-windows)
3. [Create Study Room](#create-study-room)
4. [Video Calling API](#video-calling-api)

<br />

## Starting the app in Docker<a name="starting-the-app-in-docker"></a>

To start the StudyZoom app in Docker, follow these steps:

**Step 1 - Download the code from the GitHub repository (using `git`)**

```bash
$ # Get the code
$ git clone https://github.com/jagz97/Zoom_148.git
$ cd <YOUR_BUILD_ID>
```

**Step 2 - Edit `.env` file to activate SQLite persistence**

Open the `.env` file and remove or comment out all the `DB_*` settings to activate the `SQLite` persistence. The modified `.env` file should look like this:

```txt
DEBUG=True

# Deployment SERVER address
SERVER=.zoomstudy.us

# For MySql Persistence
# DB_ENGINE=mysql            <-- REMOVE or comment for Docker
# DB_NAME=appseed_db         <-- REMOVE or comment for Docker  
# DB_HOST=localhost          <-- REMOVE or comment for Docker 
# DB_PORT=3306               <-- REMOVE or comment for Docker
# DB_USERNAME=appseed_db_usr <-- REMOVE or comment for Docker
# DB_PASS=<STRONG_PASS>      <-- REMOVE or comment for Docker

```

**Step 3 - Start the app in Docker**

```bash
$ docker-compose up --build 
```

After running this command, the StudyZoom app will start in Docker. You can access it by visiting `http://localhost:5085` in your browser.

<br />

## How to use it<a name="how-to-use-it"></a>

To use the StudyZoom app, follow the instructions below:

**Step 1 - Download the code**

```bash
$ # Get the code
$ git clone https://github.com/jagz97/Zoom_148.git
$ cd Zoom_148
```

### Set Up for Unix, MacOS<a name="set-up-for-unix-macos"></a>

**Step 2 - Install modules via `venv`**

```bash
$ virtualenv env
$ source env/bin/activate
$ pip3 install -r requirements.txt
```

**Step 3 - Set Up the Database**

```bash
$ python manage.py makemigrations
$ python manage.py migrate
```

**Step 4 - Start the app**

```bash
$ python manage.py runserver
```

After running this command, the app will start running at `http://127.0.0.1:8000/`.

### Set Up for Windows<a name="set-up-for-windows"></a>

**Step 2 - Install modules via `venv` (Windows)**

```bash
$ virtualenv env
$ .\env\Scripts\activate
$ pip3 install -r requirements.txt
```

**Step 3 - Set Up the Database**

```bash
$ python manage.py makemigrations
$ python manage.py migrate
```

**Step 4 - Start the app**

```bash
$ python manage.py runserver
```

After running this command, the app runs at `http://127.0.0.1:8000/`.

##  Create Study Room

By default, the app land on the create Study Group Page: 
<img width="740" alt="image" src="https://github.com/jagz97/Zoom_148/assets/68725607/10b1c140-672b-463e-8ae7-5956e6dc7d89">


- Create the Room Name for study group and share it with others for them to join
  **Note If you are joining a study group created by another user, enter that room name in this feild
- Enter your name to be displayed as your username
- Press join stream button and it will take you to the video chat room where you collaborate with your friends, classmates or group to study.


<br />

##  Video Calling API

The app uses Agora API for real-time video calling that provides VOIP services for real time video calling which uses RTN protocol.

The app uses simple workflow where the javascript client creates a token and the Dnajno server makes the token request and send it to the Aroga API for verificaiton and once the secure connection is established it lets users connect to the video calling services in real time.

<img width="950" alt="image" src="https://github.com/jagz97/Zoom_148/assets/68725607/05ce8936-39df-4611-9ede-fb8f3cda72e4">

The example below provides how the server creates a token for javascript client, sends json respone to aroga API.
(updated)
```py
def getToken(request):
    appId = '123e5eb9bd184d4187091c6b3a45cf89'
    appCertificate = '8a7a9fc2e8ab45a0bf3a5daf28439a53'
    channelName = request.GET.get('channel')
    uid = random.randint(1, 230)
    expirationTimeInSeconds = 3600*24
    currentTimeStamp = time.time()
    privilegeExpiredTs = currentTimeStamp + expirationTimeInSeconds
    role = 1

    token = RtcTokenBuilder.buildTokenWithUid(
        appId, appCertificate, channelName, uid, role, privilegeExpiredTs)
    return JsonResponse({'token': token, 'uid': uid}, safe=False)
```

<br />


CHANGELOG.md
# Change Log

## CodeBase version: 2stable.0.1 / 2022-01-15
### Improvements

- Dependencies update 
  - Django==4.0.1
- Settings update for Django 4.x
  - `New Parameter`: CSRF_TRUSTED_ORIGINS
    - [Origin header checking isn`t performed in older versions](https://docs.djangoproject.com/en/4.0/ref/settings/#csrf-trusted-origins)  



apps/static/js/streams.js
const APP_ID = '123e5eb9bd184d4187091c6b3a45cf89'
const CHANNEL = sessionStorage.getItem('room')
const TOKEN = sessionStorage.getItem('token')
let UID = Number(sessionStorage.getItem('UID'))

let NAME = sessionStorage.getItem('name')

const client = AgoraRTC.createClient({ mode: 'rtc', codec: 'vp8' })

let localTracks = []

let remoteUsers = {}

let joinAndDisplayLocalStream = async () => {
    document.getElementById('room-name').innerText = CHANNEL

    client.on('user-published', handleUserJoined)
    client.on('user-left', handleUserLeft)

    try {
        await client.join(APP_ID, CHANNEL, TOKEN, UID)
    } catch (error) {
        console.error(error)
        window.open('/', '_self')
    }

    localTracks = await AgoraRTC.createMicrophoneAndCameraTracks()

    let member = await createMember()

    let player = `<div class="video-container" id="user-container-${UID}">
                    <div class="username-wrapper"><span class="user-name">${member.name}</span></div>
                    <div class="video-player" id="user-${UID}"></div>
                    </div>`

    document.getElementById('video-streams').insertAdjacentHTML('beforeend', player)

    localTracks[ 1 ].play(`user-${UID}`)

    await client.publish([ localTracks[ 0 ], localTracks[ 1 ] ])
}

let handleUserJoined = async (user, mediaType) => {
    remoteUsers[ user.uid ] = user
    await client.subscribe(user, mediaType)

    if (mediaType === 'video') {
        let player = document.getElementById(`user-container-${user.uid}`)
        if (player != null) {
            player.remove()
        }

        let member = await getMember(user)

        player = `<div class="video-container" id="user-container-${user.uid}">
                    <div class="username-wrapper"><span class="user-name">${member.name}</span></div>
                    <div class="video-player" id="user-${user.uid}"></div>
                    </div>`
        document.getElementById('video-streams').insertAdjacentHTML('beforeend', player)
        user.videoTrack.play(`user-${user.uid}`)
    }

    if (mediaType === 'audio') {
        user.audioTrack.play()
    }
}

let handleUserLeft = async (user) => {
    delete remoteUsers[ user.uid ]
    document.getElementById(`user-container-${user.uid}`).remove()
}

let leaveAndRemoveLocalStream = async () => {
    for (let i = 0; localTracks.length > i; i++) {
        localTracks[ i ].stop()
        localTracks[ i ].close()
    }

    await client.leave()

    deleteMember()

    window.open('/', '_self')
}

// CAMERA TOGGLES
let toggleCamera = async (e) => {
    if (localTracks[ 1 ].muted) {
        await localTracks[ 1 ].setMuted(false)
        e.target.style.backgroundColor = '#fff'
    } else {
        await localTracks[ 1 ].setMuted(true)
        e.target.style.backgroundColor = 'rgb(255, 80, 80, 1)'
    }
}

// MIC TOGGLES
let toggleMic = async (e) => {
    if (localTracks[ 0 ].muted) {
        await localTracks[ 0 ].setMuted(false)
        e.target.style.backgroundColor = '#fff'
    } else {
        await localTracks[ 0 ].setMuted(true)
        e.target.style.backgroundColor = 'rgb(255, 80, 80, 1)'
    }
}

let createMember = async () => {
    let response = await fetch('/create_member/', {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json'
        },
        body: JSON.stringify({ 'name': NAME, 'room_name': CHANNEL, 'UID': UID })
    })
    let member = await response.json()
    return member
}

let getMember = async (user) => {
    let response = await fetch(`/get_member/?UID=${user.uid}&room_name=${CHANNEL}`)
    let member = await response.json()
    return member
}

let deleteMember = async () => {
    let response = await fetch('/delete_member/', {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json'
        },
        body: JSON.stringify({ 'name': NAME, 'room_name': CHANNEL, 'UID': UID })
    })
    let member = await response.json()
}

joinAndDisplayLocalStream()

// EXECUTE THE FUNCTIONS CREATED

window.addEventListener('beforeunload', deleteMember)

document.getElementById('leave-btn').addEventListener('click', leaveAndRemoveLocalStream)
document.getElementById('camera-btn').addEventListener('click', toggleCamera)
document.getElementById('mic-btn').addEventListener('click', toggleMic)


apps/home/views.py
# -*- encoding: utf-8 -*-
"""
Copyright (c) 2019 - present AppSeed.us
"""

from django import template
from django.contrib.auth.decorators import login_required
from django.http import HttpResponse, HttpResponseRedirect
from django.template import loader
from django.urls import reverse
from django.shortcuts import render
from django.http import JsonResponse
from agora_token_builder import RtcTokenBuilder
import random
import time
import json

from apps.home.models import RoomMember

from django.views.decorators.csrf import csrf_exempt

# Create your views here.

def index(request):
    return render(request, 'home/index.html')


def getToken(request):
    appId = '123e5eb9bd184d4187091c6b3a45cf89'
    appCertificate = '8a7a9fc2e8ab45a0bf3a5daf28439a53'
    channelName = request.GET.get('channel')
    uid = random.randint(1, 230)
    expirationTimeInSeconds = 3600*24
    currentTimeStamp = time.time()
    privilegeExpiredTs = currentTimeStamp + expirationTimeInSeconds
    role = 1

    # role = 1 => HOST || role = 2 => GUEST

    token = RtcTokenBuilder.buildTokenWithUid(
        appId, appCertificate, channelName, uid, role, privilegeExpiredTs)
    return JsonResponse({'token': token, 'uid': uid}, safe=False)


def main(request):
    return render(request, 'home/landing.html')


def chatroom(request):
    return render(request, 'home/chatroom.html')


@csrf_exempt
def createMember(request):
    data = json.loads(request.body)

    member, created = RoomMember.objects.get_or_create(
        name=data['name'],
        uid=data['UID'],
        room_name=data['room_name']
    )
    return JsonResponse({'name': data['name']}, safe=False)


def getMember(request):
    uid = request.GET.get('UID')
    room_name = request.GET.get('room_name')

    member = RoomMember.objects.get(
        uid=uid,
        room_name=room_name,
    )

    name = member.name
    return JsonResponse({'name': member.name}, safe=False)


@csrf_exempt
def deleteMember(request):
    data = json.loads(request.body)

    chat_member = RoomMember.objects.get(
        name=data['name'],
        uid=data['UID'],
        room_name=data['room_name'],
    )
    chat_member.delete()
    return JsonResponse('Member removed', safe=False)
